> @tmuehlbacher-bnr are any of those targan bugs actually valid tho? I had a quick look but I couldn't see anything actually wrong...

Scrolling through the action build logs. Seems I miscounted at first, thinking the last 2 entries were the 4 bugs it claims at the end because these both generated 2 warnings each.

```
[569/2124] Compiling C object plugins/devlink/libfu_plugin_devlink.a.p/fu-devlink-netlink.c.o
../../../plugins/devlink/fu-devlink-netlink.c:294:8: warning: Null pointer passed to 1st parameter expecting 'nonnull' [core.NonNullParamChecker]
  294 |                             fu_device_event_get_bytes(event, response_key, NULL);
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../plugins/devlink/fu-devlink-netlink.c:328:5: warning: Null pointer passed to 1st parameter expecting 'nonnull' [core.NonNullParamChecker]
  328 |                                 fu_device_event_set_bytes(event, response_key, response_data);
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

```
[674/2124] Compiling C object plugins/ilitek-its/libfu_plugin_ilitek_its.a.p/fu-ilitek-its-device.c.o
../../../plugins/ilitek-its/fu-ilitek-its-device.c:538:3: warning: Overwriting already-set GError []
  507 |                 g_set_error(error,
      |                 ^~~~~~~~~~~~~~~~~~
  508 |                             FWUPD_ERROR,
      |                             ~~~~~~~~~~~~
  509 |                             FWUPD_ERROR_INVALID_FILE,
      |                             ~~~~~~~~~~~~~~~~~~~~~~~~~
  510 |                             "no HID_PHYS in %s",
      |                             ~~~~~~~~~~~~~~~~~~~~
  511 |                             fu_udev_device_get_sysfs_path(parent));
      |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.

```

These are the one's I thought were all it found at first and those seem not relevant or someone else's problem:

```
[1250/2124] Generating libfwupd/Fwupd-2.0.gir with a custom command (wrapped by meson to set env)
/github/workspace/build/meson-private/tmp344o8r8g/tmp-introspectic6sic8v/Fwupd-2.0.c:725:15: warning: Overwriting already-set GError []
  154 |               g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_FAILED,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  155 |                            "Invalid GType function: %s", function);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/github/workspace/build/meson-private/tmp344o8r8g/tmp-introspectic6sic8v/Fwupd-2.0.c:746:15: warning: Overwriting already-set GError []
  179 |               g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_FAILED,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  180 |                            "Invalid error quark function: %s", function);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

```
[1563/2124] Generating libfwupdplugin/FwupdPlugin-1.0.gir with a custom command (wrapped by meson to set env)
/github/workspace/build/meson-private/tmp344o8r8g/tmp-introspect8o_tz11r/FwupdPlugin-1.0.c:725:15: warning: Overwriting already-set GError []
  154 |               g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_FAILED,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  155 |                            "Invalid GType function: %s", function);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/github/workspace/build/meson-private/tmp344o8r8g/tmp-introspect8o_tz11r/FwupdPlugin-1.0.c:746:15: warning: Overwriting already-set GError []
  179 |               g_set_error (error, G_FILE_ERROR, G_FILE_ERROR_FAILED,
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  180 |                            "Invalid error quark function: %s", function);
      |                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```

--- a/contrib/ci/Dockerfile-debian-tartan.in
+++ b/contrib/ci/Dockerfile-debian-tartan.in
@@ -4,6 +4,7 @@ ENV CI_NETWORK=true
 RUN echo fubar > /etc/machine-id
 RUN set -euxo pipefail; \
     export DEBIAN_FRONTEND=noninteractive; \
+%%%SETUP%%% \
     apt-get update; \
     apt-get install --yes --no-install-recommends \
     build-essential \

--- a/contrib/ci/Dockerfile-debian.in
+++ b/contrib/ci/Dockerfile-debian.in
@@ -4,9 +4,7 @@ ENV CI_NETWORK=true
 RUN echo fubar > /etc/machine-id
 RUN set -euxo pipefail; \
     export DEBIAN_FRONTEND=noninteractive; \
-    sed -i 's|Types: deb|Types: deb deb-src|' /etc/apt/sources.list.d/debian.sources; \
-    dpkg --add-architecture i386; \
-    dpkg --add-architecture s390x; \
+%%%SETUP%%% \
     apt-get update; \
     apt-get install --yes --no-install-recommends \
     build-essential \

--- a/contrib/ci/README.md
+++ b/contrib/ci/README.md
@@ -110,7 +110,7 @@ The following builds are performed for every commit or pull request:
 ## Adding a new target
 
 Dockerfiles are generated dynamically by the python script `generate_docker.py`.
-The python script will recognize the environment variable `TARGET_DISTRO` to determine what target to generate a Dockerfile for.
+The python script will recognize the environment variable `DISTRO` to determine what target to generate a Dockerfile for.
 
 ### dependencies.xml
 

--- a/contrib/ci/generate_docker.py
+++ b/contrib/ci/generate_docker.py
@@ -9,6 +9,7 @@ import os
 import shutil
 import subprocess
 import sys
+from pathlib import Path
 
 from fwupd_setup_helpers import parse_dependencies
 
@@ -43,37 +44,32 @@ def get_container_cmd():
 directory = os.path.dirname(sys.argv[0])
 DISTRO = getenv_unwrap("DISTRO")
 ARCH = getenv_unwrap("ARCH")
-match os.getenv("VARIANT"):
-    case str(s) if s.startswith("cross-"):
-        CROSS = s.removeprefix("cross-")
-        VARIANT = None
-    case str(s):
-        CROSS = None
-        VARIANT = s
-    case _:
-        CROSS = None
-        VARIANT = None
+VARIANT = os.getenv("VARIANT")
+CROSS = c if (c := str(VARIANT).removeprefix("cross-")) != VARIANT else None
 
 
-if VARIANT:
-    template_file = os.path.join(directory, f"Dockerfile-{DISTRO}-{VARIANT}.in")
-else:
-    template_file = os.path.join(directory, f"Dockerfile-{DISTRO}.in")
-if not os.path.exists(template_file):
-    print(f"Missing input file {template_file} for {DISTRO}")
-    sys.exit(1)
-
+# find first existing
+dockerfiles = [
+    Path(directory) / f"Dockerfile-{DISTRO}-{VARIANT}.in",
+    Path(directory) / f"Dockerfile-{DISTRO}.in",
+]
+try:
+    template_file = next(p for p in dockerfiles if p.exists())
+except StopIteration:
+    raise FileNotFoundError("Missing template Dockerfile for {DISTRO}") from None
 with open(template_file) as file:
-    template = file.read()
+    content = file.read()
 
 
+# special cases
 match (DISTRO, VARIANT):
     case ("debian", "i386"):
-        template = template.replace("FROM debian:testing", "FROM i386/debian:testing")
+        content = content.replace("FROM debian:testing", "FROM i386/debian:testing")
     case ("debian", "tartan"):
-        template = template.replace("FROM debian:testing", "FROM debian:unstable")
+        content = content.replace("FROM debian:testing", "FROM debian:unstable")
 
 
+# insert dependencies to install
 if CROSS:
     deps = parse_dependencies(DISTRO, ARCH_TO_DEPS_MAP[CROSS], False, cross=True)
     deps += [f"crossbuild-essential-{CROSS}"]
@@ -82,8 +78,8 @@ else:
 deps = sorted(set(deps))
 deps = [f"    {i}" for i in deps]
 deps = " \\\n".join(deps)
+content = content.replace("%%%DEPENDENCIES%%%", deps)
 
-content = template.replace("%%%DEPENDENCIES%%%", deps)
 
 with open("Dockerfile", "w") as file:
     file.write(content)

--- a/contrib/ci/generate_docker.py
+++ b/contrib/ci/generate_docker.py
@@ -69,6 +69,16 @@ match (DISTRO, VARIANT):
         content = content.replace("FROM debian:testing", "FROM debian:unstable")
 
 
+# insert commands to prepare cross compile
+if CROSS:
+    cross_setup = f"""\
+    sed -i 's|Types: deb|Types: deb deb-src|' /etc/apt/sources.list.d/debian.sources; \\
+    dpkg --add-architecture {CROSS};"""
+else:
+    cross_setup = "    "
+content = content.replace("%%%SETUP%%%", cross_setup)
+
+
 # insert dependencies to install
 if CROSS:
     deps = parse_dependencies(DISTRO, ARCH_TO_DEPS_MAP[CROSS], False, cross=True)

